### ISSUE TITLE
Project is dead

### ISSUE THREAD

**Issue by kai-nashi** [vuelidate/vuelidate#1163]
Last release about 6 months ago. No active contributors, no community merge requests, no bug fixes. It still work but have a lot of problems. Reasonable to not use it.

**Comment by galaxyblur** [comment]
Know of any alternatives?

**Comment by JoeHart** [comment]
Vee validate looks like a good alternative:
https://github.com/logaretm/vee-validate

**Comment by dobromir-hristov** [comment]
Unfortunately it's a different approach at the same idea.

If you want data driven vilidation, there ain't much else.

The project is not dead, but it is no longer actively developed.

**Comment by galaxyblur** [comment]
Dead, in this context, means no longer maintained. Which is important for bug fixes and security updates, at the very least.

And that's not a criticism—thanks to you and to others for this open-source library, I've been able to do some great work. But if it's now simply legacy code, it quickly becomes tech debt. Which means looking for an alternative or contributing to this project. I see some open PRs, which further indicate that there is no intent to maintain this project, so I'm inclined to look elsewhere.

**Comment by dobromir-hristov** [comment]
No prob, by all means try veevalidate. It's a great project.

**Comment by Nikakto** [comment]
2.0 has broken most of killer features that was realized by @Frizi.

Nested validation require active component and agregate all errors to one array.
New forEarch is broken and does not providing api to access $v for item in array.

VeeValidate is not so flexible like vuelidate 0.x and code hard to read. 
Vuelidate 2.x has a lot of bugs.
There are no good alternatives.

**Comment by dawidstefaniak** [comment]
If you are looking for data-based validation instead of component-based vee-validate, you can take a look on joi: https://joi.dev/ or yup https://github.com/jquense/yup. You can use the same validators for BE and FE if your backend is written in JS.

**Comment by JoeHart** [comment]
Looking around at other options react-hook-forms just released their core lib that's framework agnostic and might be a good basis to build form functionality into a vue app. Might have a hack around with it today: https://github.com/react-hook-form/core

**Comment by galaxyblur** [comment]
I was going to mention something similar to joi and yup: zod. I realized I already have schemas for my data so I can pretty easily mimic what vuelidate was doing.

**Comment by milelazar** [comment]
You can try https://formkit.com/. Still in beta but it looks good.

**Comment by metalsadman** [comment]
sad, this was the best approach with vue validation by far imo.

**Comment by Brunnn** [comment]
Please, consider putting a notice on the project/website of this package if its no longer maintained, I've chosen this package 7 months ago because it looked reasonably active and wide used. But with the currently state of vue 3 updates and changes this library just became a mess of awkward behavior, confusing and not updated documentation.

At the current state i strongly disencourage anyone to use this lib for Vue 3, specially if you're using typescript.

**Comment by nolde** [comment]
Any specific reason why it is no longer supported, @dobromir-hristov ?

Is it related to funding?

**Comment by pczarn** [comment]
I may pick up the project as a developer and maintainer within a few months.

I helped the vuelidate creators make vuelidate 6 years back.

**Comment by dobromir-hristov** [comment]
Reason is I no longer use vuelidate, I unfortunately don't have time to learn proper TS to migrate it over, in order to fix all the annoying issues TS users have.

I am working from 8 to 9 at my current job, for various timezone and other reasons, which pays more than I can ever make with oss. 

The project as it stands now just needs a freshen up, some minor cleanup and making sure it works with the broader ecosystem like Nuxt.

The API is mostly fine, with some quirks and edge cases that came from trying to support both old and new API.

If anyone wants to help out and create some PRs, by all means. I will make sure to review in a timely manner.

**Comment by nolde** [comment]
Thanks for letting us know, @dobromir-hristov .

Yeah, that's quite fair. If it was funding only, I could try to get my company to add to the jar. We already sponsor vuetify, might as well contribute to another project that we heavily use.

All the other options are not feasible for us, exactly due to being the only option that is pure data driven validation; it is so much cleaner and straightforward. 

I already made a change to support non-object lists with forEach, I might PR that in soon.

**Comment by dobromir-hristov** [comment]
I have started refactoring the library build process, so we don't get issues with latest Vite.

I merged a few PRs and will make sure it works with Nuxt. 

Again I am very limited on time as I have to do this before my family wakes up on weekends...

**Comment by Brunnn** [comment]
Our team decided to move our schema validations to the [Zod Library](https://zod.dev/), it is very flexible and it has Typescript-First principles. It doesn't have any built-in support for vue reactiveness, but that can be easlily implemented. Its a good schema validation library for people using vue + typescript.

EDIT: We found a package to implement Zod Alongside Vee-validade, it's working flawlessly, see: https://vee-validate.logaretm.com/v4/integrations/zod-schema-validation/

**Comment by metalsadman** [comment]
> Our team decided to move our schema validations to the [Zod Library](https://zod.dev/), it is very flexible and it has Typescript-First principles. It doesn't have any built-in support for vue reactiveness, but that can be easlily implemented. Its a good schema validation library for people using vue + typescript.

You can also use it with VeeValidate as well.

**Comment by carlpage** [comment]
Our team is using Vuetify as the component library and decided to switch to its built-in form validation. We have had a lot of success with it since the switch!

**Comment by martinszeltins** [comment]
I love this project, it makes working with forms so easy. Hope it is not going to die.

**Comment by metalsadman** [comment]
> I love this project, it makes working with forms so easy. Hope it is not going to die.

same, I hope someone fork this up with better ts support, this was always my goto vue validation lib.

**Comment by dobromir-hristov** [comment]
I have been working more and more with TS and a TS rewrite is not unheard of.

I can deff review work if someone wants to start or at least to refactor and cleanup the current types.

A full rewrite is something I have wanted to do for about a year...

I will need to learn how to properly generate a package with the types and modules and all the stuff... Releasing packages is never simple...

**Comment by martinszeltins** [comment]
@dobromir-hristov Since Vuelidate is such an amazing library, isn't it possible to keep working on the library as it is right now and improve / maintain it without a TS rewrite? Perhaps TS rewrite could be done without abandoning the previous version?

**Comment by kirkbushell** [comment]
> Unfortunately it's a different approach at the same idea.
> 
> If you want data driven vilidation, there ain't much else.
> 
> The project is not dead, but it is no longer actively developed.

Oh damn, had no idea. Someone needs to fork this then because this is easily the best validation library for Vue.

**Comment by jacobg** [comment]
Any forks with Typescript working. PrimeVue is giving the following typescript error when setting v-model to $model:
```vue
 error TS2322: Type 'Ref<string>' is not assignable to type 'Nullable<string>'.

<InputText v-model.trim="v$.email.$model" v-bind="slotProps" :disabled="processing" />
```

**Comment by renatodeleao** [comment]
Hey everyone! Just trying to get the pulse on this library current usage in Vue the ecosystem.

It's now clear that this project is abandoned and previous maintainers don't have bandwidth or desire to maintain it — **which is totally fine and a perfect valid option that we should only respect and thanks for the awesome work so far.**

Now, I still use it, and i'm heavily invested on it in some projects and as such have been spending a lot of time finding workarounds for things that should work out-of-the-box. If I was starting a new project today, I would choose `vee-validate` as nowadays is also capable of pure _model based validation_, has a few extra goodies and it's well maintained.

That being said, `vuelidate` is still awesome! I would like to volunteer to take care of some of the issues, but I would like to know if I would be on this journey alone: one-man-show projects tend to die very fast, so no point in even starting just to become overwhelmed with issues and give up a few weeks later. I have 0 interest in fixing Typescript wizardry issues, so type wizards are also welcomed to join on board.

So if you have interest, comment below and if we have a good-enough group of people (I'll put the min treshold of at least 3), let's mention the maintainers to discuss the possibility of being added to the team or do a fork.

Cheers ✌️

**Comment by martinszeltins** [comment]
I support the idea of @renatodeleao 100%. I don't have the time or know-how myself but have been thinking someone PLEASE maintain this awesome project.

**Comment by VladBrok** [comment]
Is there an alternative for vuelidate but for React?

**Comment by shentao** [comment]
@renatodeleao I'm happy to welcome new maintainers if anyone is up for it.

**Comment by pczarn** [comment]
@shentao Hi, Piotr here. I don't have the capacity to maintain anything this year. I've been maintaining only one project, Rust's bit-vec.

**Comment by victorgarciaesgi** [comment]
Hey all, as a Vuelidate lover it's sad to see it die! 
For a while I was working on a successor to Vuelidate with Typescript first and Zod support that I called `Regle` !
The project is almost ready (docs and e2e tests to finish)
Don't hesitate to try it here! 😁 https://github.com/victorgarciaesgi/regle

**Comment by martinszeltins** [comment]
> Hey all, as a Vuelidate lover it's sad to see it die! For a while I was working on a successor to Vuelidate with Typescript first and Zod support that I called `Regle` ! The project is almost ready (docs and e2e tests to finish) Don't hesitate to try it here! 😁 [victorgarciaesgi/regle](https://github.com/victorgarciaesgi/regle)

@victorgarciaesgi  That is really nice to see. I just have one question, why is nested collection validation not planned? Can you explain? Also, does "autoDirty" work with collections now? The last release was in April, why is that?

**What is not planned or reimplemented differently from Vuelidate**
- Validation groups
- Nested validation collection

**Comment by shentao** [comment]
Looks promising @victorgarciaesgi! Fingers crossed!

If I may have one suggestion, I'd say you might want to implement `nested validation collection` in some way. IMO, it's one of Vuelidate's best features.

**Comment by victorgarciaesgi** [comment]
Maybe I was not clear in the readme, collections validations are here with $each, but not nested component validation (as I can't infer Typescript types from it)

**Comment by kirkbushell** [comment]
It'd be sad for this project to die. I'm not a typescript master by any stretch of the imagination, but am entertaining the idea of forking and maintaining in order to improve my skills there. 

I really love this library, I think it's extremely powerful but just needs a little love.

**Comment by victorgarciaesgi** [comment]
@martinszeltins I have a full time job so finding time to work on it is hard. The core is ready since April, I still need to find time for finishing tests and docs

**Comment by victorgarciaesgi** [comment]
@shentao Yeah for nested collection validation It's more a personal choice, I prefer the validation to be fully data-based, not component based. Also the hard part to develop is Typescript inference for it, which is a core part of Regle

**Comment by martinszeltins** [comment]
> Maybe I was not clear in the readme, collections validations are here with $each, but not nested component validation (as I can't infer Typescript types from it)

@victorgarciaesgi yes, the readme is wrong then because it says that `Nested validation collection` is not planned. I don't need nested component validation but I do need nested collection validation. Also, Vuelidate had a bug where autoDirty would not work for collections. Does `regle` support autoDirty for each collection item field?

I understand that most of us have full-time jobs but would you have time to do issues and pull requests if they were submitted to the project? I would love to try it out, test it and report any bugs but I want to know that the new project is still maintained and the issues will be addressed.

**Comment by victorgarciaesgi** [comment]
Yep I know it's hard to find time, that's why it's not really released yet. 
It's not new to me, I maintain for a long time now a Nuxt module and it's still active https://github.com/victorgarciaesgi/nuxt-typed-router
Thanks for the recommendations!

**Comment by martinszeltins** [comment]
@victorgarciaesgi currently it has this warning that it should not be used in production. But since last release was back in April, when do you think we can expect a stable, production-ready release?

> ⚠️ Project is still in development, do not use it in production API or function names can still change

**Comment by shentao** [comment]
> @shentao Yeah for nested collection validation It's more a personal choice, I prefer the validation to be fully data-based, not component based. Also the hard part to develop is Typescript inference for it, which is a core part of Regle

@victorgarciaesgi The use case where it shines is when you are working with large, user-built forms (based on a schema) that can be nested several times etc. Oftentimes, you want each field to manage and validate itself, especially if the field's schema is dynamic (as in, its configuration changes based on other fields in the form and/or outside data). In such situations, you don't want to mutate the top-level schema, as it would trigger a large downstream effect. Instead, you likely want to keep the adjustments local. With `nested validation collection` it's so much easier to have the parent component be aware that there are errors inside, without really having to be aware of all the nested fields. So it's still fully data-driven. But yeah, that's a very specific use case I work with and why Vuelidate introduced such a feature. Without it (old Vuelidate), the code was a mess when you wanted to keep track of different validations spread around the form (and nested forms).

**Comment by jacobg** [comment]
Just curious, why isn't VeeValidate an acceptable alternative? It seems years ago, the major difference between Vuelidate and VeeValidate was that Vuelidate centered around validating state whereas VeeValidate centered around validating forms. But it seems that since then VeeValidate has made it easier to validate state and has full typescript support. It also has longevity and community backing.

**Comment by shentao** [comment]
It is an acceptable alternative. I use Vuelidate because it provides the solutions I need at my current job.

**Comment by jacobg** [comment]
I started playing around with VeeValidate, and really came to an appreciation of how easy it is to work with Vuelidate. The benefits I find of Vuelidate are:
- add in validation without polluting original data model or needing to copy values back and forth, i.e., better separation of concerns
- unified vue reactivity of everything (validators, field names, messages, etc), which you don't get in VeeValidate especially with adding in zod schemas
- simple and much less boilerplate

That said, I think there are some reactivity bugs in Vuelidate in certain scenarios, and its Typescript needs attention. Then there is the lack of community support.

So I may try to develop patterns of using VeeValidate to get better reactivity and that minimize boilerplate. So far, their docs are a bit overwhelming.

**Comment by martinszeltins** [comment]
> I started playing around with VeeValidate, and really came to an appreciation of how easy it is to work with Vuelidate. The benefits I find of Vuelidate are:
> 
> * add in validation without polluting original data model or needing to copy values back and forth, i.e., better separation of concerns
> * unified vue reactivity of everything (validators, field names, messages, etc), which you don't get in VeeValidate especially with adding in zod schemas
> * simple and much less boilerplate
> 
> That said, I think there are some reactivity bugs in Vuelidate in certain scenarios, and its Typescript needs attention. Then there is the lack of community support.
> 
> So I may try to develop patterns of using VeeValidate to get better reactivity and that minimize boilerplate. So far, their docs are a bit overwhelming.

@jacobg Look into regle, it is also a data model driven validation like Vuelidate - https://regle.vercel.app/

**Comment by metalsadman** [comment]
> > I started playing around with VeeValidate, and really came to an appreciation of how easy it is to work with Vuelidate. The benefits I find of Vuelidate are:
> > 
> > * add in validation without polluting original data model or needing to copy values back and forth, i.e., better separation of concerns
> > * unified vue reactivity of everything (validators, field names, messages, etc), which you don't get in VeeValidate especially with adding in zod schemas
> > * simple and much less boilerplate
> > 
> > That said, I think there are some reactivity bugs in Vuelidate in certain scenarios, and its Typescript needs attention. Then there is the lack of community support.
> > So I may try to develop patterns of using VeeValidate to get better reactivity and that minimize boilerplate. So far, their docs are a bit overwhelming.
> 
> @jacobg Look into regle, it is also a data model driven validation like Vuelidate - https://regle.vercel.app/

Looks promising and zod integration too, i hope the author integrates yup there's an issue I have with zod about superRefine.

**Comment by victorgarciaesgi** [comment]
@metalsadman hey! I'm the author, yup can totally be in the roadmap 👌🏻. I wanted to do Valibot before as it's newer

**Comment by metalsadman** [comment]
> @metalsadman hey! I'm the author, yup can totally be in the roadmap 👌🏻. I wanted to do Valibot before as it's newer

Cool! Will keep an eye.

**Comment by jacobg** [comment]
@victorgarciaesgi  How is regle different than veevalidate?

**Comment by victorgarciaesgi** [comment]
It's exactly like you describe here:
> - add in validation without polluting original data model or needing to copy values back and forth, i.e., better separation of concerns
> - unified vue reactivity of everything (validators, field names, messages, etc), which you don't get in VeeValidate especially with adding in zod schemas
> - simple and much less boilerplate

I took what made Vuelidate good, making it type safe, adding features and compatible with validation libraries like `zod` (while staying reactive)

**Comment by nickdnk** [comment]
> It's exactly like you describe here:
> 
> > * add in validation without polluting original data model or needing to copy values back and forth, i.e., better separation of concerns
> > * unified vue reactivity of everything (validators, field names, messages, etc), which you don't get in VeeValidate especially with adding in zod schemas
> > * simple and much less boilerplate
> 
> I took what made Vuelidate good, making it type safe, adding features and compatible with validation libraries like `zod` (while staying reactive)

Hey. I just found some weird bugs in Vuelidate and ended up here. I totally did not notice that we were working with a practically dead package, which is concerning considering how many lines of vuelidate code we have deployed. Anyway, how "production ready" would you consider your project to be, @victorgarciaesgi? It looks pretty good on the docs and I don't have a problem with bleeding edge if the creator says it works, but of course, if you kick the bucket tomorrow we're back to square one I guess. How's your health? Hehe. Jokes aside, I'm trying to make this migration as painless as possible as it's *a lot* of components we have to migrate, so something closer to Vuelidate (the alternative being VeeValidate it seems) would be better for us.

**Comment by victorgarciaesgi** [comment]
@nickdnk It's already production ready! Using it in prod right now at my company and will migrate existing Vuelidate forms in the long run. 
Migrations are easy as I tried to keep the same API while improving some areas, and as it's 100% type safe it should spare you some struggle 😄

**Comment by nickdnk** [comment]
> @nickdnk It's already production ready! Using it in prod right now at my company and will migrate existing Vuelidate forms in the long run. 
> 
> Migrations are easy as I tried to keep the same API while improving some areas, and as it's 100% type safe it should spare you some struggle 😄 

Excellent news. Hopefully we can get the majority of this done with some clever regex-based search/replace then. I will try it out after new year's.

**Comment by victorgarciaesgi** [comment]
I'll prepare a migration guide in the docs, that could help!
Next week I should release the `1.0` and make an announcement somewhere (didn't made one yet)

**Comment by victorgarciaesgi** [comment]
https://regle.vercel.app/introduction/migrate-from-vuelidate

**Comment by dixhuit** [comment]
I'm still not sure what my best migration option is as the maintainer of a large codebase that relies heavily on Vuelidate's nested component validation. What are others in this situation planning to do?

**Comment by nickdnk** [comment]
> I'm still not sure what my best migration option is as the maintainer of a large codebase that relies heavily on Vuelidate's nested component validation. What are others in this situation planning to do?

I'll go with Regle and then if that project should die, I'll do a larger migration to VeeValidate, which isn't feasible right now.

**Comment by dixhuit** [comment]
> > I'm still not sure what my best migration option is as the maintainer of a large codebase that relies heavily on Vuelidate's nested component validation. What are others in this situation planning to do?
> 
> I'll go with Regle and then if that project should die, I'll do a larger migration to VeeValidate, which isn't feasible right now.

I thought Regle didn't support nested _component_ validation? (it's discussed further up in this issue, search the page for "nest")

**Comment by nickdnk** [comment]
> > > I'm still not sure what my best migration option is as the maintainer of a large codebase that relies heavily on Vuelidate's nested component validation. What are others in this situation planning to do?
> > 
> > 
> > I'll go with Regle and then if that project should die, I'll do a larger migration to VeeValidate, which isn't feasible right now.
> 
> I thought Regle didn't support nested _component_ validation? (it's discussed further up in this issue, search the page for "nest")

Ah. I just pass in a validation rule as a property to a child component if I want to validate nested components. I don't think we use nested component validation at all with Vuelidate because it tended to either not work of perform bad. Not sure though. Or maybe the performance issue was only for collections.

**Comment by dixhuit** [comment]
> > > > I'm still not sure what my best migration option is as the maintainer of a large codebase that relies heavily on Vuelidate's nested component validation. What are others in this situation planning to do?
> > > 
> > > 
> > > I'll go with Regle and then if that project should die, I'll do a larger migration to VeeValidate, which isn't feasible right now.
> > 
> > 
> > I thought Regle didn't support nested _component_ validation? (it's discussed further up in this issue, search the page for "nest")
> 
> Ah. I just pass in a validation rule as a property to a child component if I want to validate nested components. I don't think we use nested component validation at all with Vuelidate because it tended to either not work of perform bad. Not sure though. Or maybe the performance issue was only for collections.

Yeah, I can't do that as my use case requires each component to define its own validation (which is really handy when you have lots of different components and any of them could potentially end up providing groups of inputs within your form). This was always the killer feature in Vuelidate for me - nothing seems to offer a substitute.

**Comment by victorgarciaesgi** [comment]
Yeah there is no nested component validation. 
It's by choice because my approach was to be type-safe first and declarative, so the nested component collection was the opposite of what I wanted to do.
The support for collections and `$each` is upgraded from Vuelidate. It tracks item correctly (except primitives, JS limitation..)

**Comment by jacobg** [comment]
> Yeah there is no nested component validation. It's by choice because my approach was to be type-safe first and declarative, so the nested component collection was the opposite of what I wanted to do. The support for collections and `$each` is upgraded from Vuelidate. It tracks item correctly (except primitives, JS limitation..)

I just tried regle today for the first time, and so far so good! Thanks for making this great library. The docs look very thorough and well thought out!

I never used vuelidate nested component validation, because I had only refactored my Vue 2 app with vuelidate and never added new functionality with vuelidate. But just curious why it's not possible to do typescript nested component validation as documented here?
https://vuejs.org/guide/typescript/composition-api#typing-component-template-refs

**Comment by dobromir-hristov** [comment]
Nested component validation could be done via some userland collector.

You would need to be explicit what emits and what collects, and for typesafety, you would have to do some manual type providing. Similar to how Zod can export schemas, you could grab something like that from the validator instance or something. A bit error prone, but I don't see a better way atm

**Comment by dobromir-hristov** [comment]
@jacobg you don't always have direct ref binding to child forms.

Often they are inside other components.

Even if you had, I think it's just static types like exported methods, refs etc. Not sure if Vue  compiler will manage to dig through your validation rules to generate the schema.

**Comment by jacobg** [comment]
> @jacobg you don't always have direct ref binding to child forms.
> 
> Often they are inside other components.
> 
> Even if you had, I think it's just static types like exported methods, refs etc. Not sure if Vue compiler will manage to dig through your validation rules to generate the schema.

@dobromir-hristov Does vuelidate recursively look through children for v$ even through direct descendants not using it?

**Comment by dobromir-hristov** [comment]
Currently yes, though IMHO that is not a great solution and at the time it felt magic, now I really don't like it😅 

You have very little control over what ends up in your rulebook so to speak.

**Comment by victorgarciaesgi** [comment]
Yep template refs are a way, you will have the expose the correct binding for the parent component to collect.
So you lose the main argument that is to collect automatically children validations. 
If you have to declare templateRefs for each components, you better of use props at this point.

**Comment by jacobg** [comment]
What I did from Vue 2 implementation was have the child emit an `@validity-changed` event, and the parent would add a validator that the child is valid. But the exact child errors were encapsulated in the child, and displayed by it. For the parent, it's just useful to disable the save button and maybe display an error message next to the save button saying `** Fix errors above before saving`.

**Comment by dixhuit** [comment]
Agreed, the top-level parent doesn't need to know all of the gory details, just that a child component somewhere is invalid and that that is enough to render the entire form state invalid. That child component can then be responsible for its own validation and still be declarative & type safe. Would this approach not give us the best of both worlds?

**Comment by victorgarciaesgi** [comment]
I see what you mean. If I try to implement this, I will not mix it up with other properties and maybe provide a `$children` additional property with just `$valid, $invalid, $pending` common properties

**Comment by dixhuit** [comment]
> I see what you mean. If I try to implement this, I will not mix it up with other properties and maybe provide a `$children` additional property with just `$valid, $invalid, $pending` common properties

Great! I've created the following Regle issue as a follow up: https://github.com/victorgarciaesgi/regle/issues/40

**Comment by shentao** [comment]
Great conversations here! Makes me actually miss working on OSS. Maybe something to do in 2025? 🤔

If you guys feel like contributing to the project, let me know. 

About the limiting of error data reported to the parents – I don't think this would be fully sufficient for cases where perhaps you want to show what exact errors some of the nested fields have (for example, next to a submit button).

**Comment by dixhuit** [comment]
> About the limiting of error data reported to the parents – I don't think this would be fully sufficient for cases where perhaps you want to show what exact errors some of the nested fields have (for example, next to a submit button).

This is a limitation for sure but in my mind it's kinda mitigated by the fact that each component would be responsible for defining and displaying its own messages (e.g. displayed near each input rather than grouped in a single location), much like how Vuelidate does now. I think it's about compromise, it sounds like the current direction of Regle, being one of type-safety through up-front declaration, makes having detailed collection of error data in a single place (e.g. `$r` / `$v`) difficult. I'm no TS expert though so I'm just basing these assumptions on what I've read in this thread!

**Comment by victorgarciaesgi** [comment]
Yep exactly, the type safety, and even more so the type inference are a big challenge for this feature.
I do not know the full extent of the use cases as I never used this feature. I will have to experiment things :D

**Comment by shentao** [comment]
> This is a limitation for sure but in my mind it's kinda mitigated by the fact that each component would be responsible for defining and displaying its own messages (e.g. displayed near each input rather than grouped in a single location), much like how Vuelidate does now. I think it's about compromise, it sounds like the current direction of Regle, being one of type-safety through up-front declaration, makes having detailed collection of error data in a single place (e.g. `$r` / `$v`) difficult. I'm no TS expert though so I'm just basing these assumptions on what I've read in this thread!

Yeah, type safety can be challenging in such cases. But I think it should be doable, since right now it puts the results within the existing results collection that has a more uniform type. 

As for the feature itself, as stated before, I believe it's a core feature that differentiates Vuelidate from other solutions. It's also a critical feature for my own use case (huge, multiply nested user-created schema-based forms). The form fields display their own validation results, but we also need a way to keep a list of results for a "card" container of a nested form and the top level submit/save buttons with navigation links to specific fields. Otherwise it's hard to find the field that is failing validation. Additionally, each field reacts to the form's value and dynamically updates its own value and/or configuration, which includes validation rules. Having this at top level would be very inefficient performance wise.

I wonder if maybe building something like this (a formvuelate-like solution) in a video course format is something that folks would be interested in. Thoughts?

**Comment by shentao** [comment]
Also, saw the new documentation for Regle and it looks neat! Nice work @victorgarciaesgi!

Happy new year guys!

**Comment by whiplashwebb** [comment]
this is a very helpful discussion, both for understanding vuelidate's current status and migration options. I'm also in the situation where I've been happily using Vuelidate for years and didn't realize it wasn't maintained until recently. Much love to everyone who created it, it's the gold standard of vue validation imho, I hope you ship an update but I appreciate all you've given us already regardless. it's really nice to find this much life at the end of a thread with a depressing name. @victorgarciaesgi i'm 100% checking out regle, I love how you even made a migration guide!

**Comment by dixhuit** [comment]
@whiplashwebb Makes a good point re the issue title, perhaps @kai-nashi or @dobromir-hristov  can update it to offer a bit more hope to Vuelidate users looking for answers? Perhaps:

_Project is dead. "Regle" is a possible replacement?_

Or would it be more appropriate to start a new issue specifically to signpost possible alternatives?

**Comment by shentao** [comment]
Thanks @whiplashwebb! It means a lot to me, to hear that!

I wouldn't mind pinning a new issue that could include a list of possible alternatives with a short comparison of how they differ from Vuelidate. cc @victorgarciaesgi

**Comment by victorgarciaesgi** [comment]
@shentao Can I just add that I used Vuelidate for 8 years and it's was always my go-to solution for forms (complex and simple), and by far the best API to handle forms (thanks to Vue reactivity too).
Couldn't think  of using another library, but sadly I know it's hard to maintain (not on that scale yet), so I wanted to continue the legacy and make the model-based validation new again! 

Thanks again for having such a great solution from the start ❤ 

If it's ok with you I can help make a list of comparisons.

But I just want to warn, I don't consider `Regle` in a 100% ready state (it's not in `1.0` yet), i'm nearly there, but I have still feature planned to really make it a full replacement

**Comment by martinszeltins** [comment]
> I wouldn't mind pinning a new issue that could include a list of possible alternatives with a short comparison of how they differ from Vuelidate.

I don't think there are any real alternatives of **data-driven** validation libraries to Vuelidate other than Regle. There are of course multiple DOM / Component driven libraries like VeeValidate, FormKit but their model is very different from Vuelidate and Regle.

There are basically 2 types of form validation libraries - element driven and data driven. As far as I know Vuelidate and Regle are the only data / model driven ones.

Vue by nature is also data-driven so Vuelidate/Regle fit in perfectly in this model.

**Comment by whiplashwebb** [comment]
@martinszeltins i'm 100% with you, I use vuelidate specifically because i want data driven validation. I'm sure almost everyone on this thread is in the same boat. but even though i know exact whatly i want I still have to look at the alternatives before I make a big library decision like this. i'm running vuelidate in prod. one of the really helpful things about this thread is hearing the arguments for the alternatives, I feel much better about sticking with the data-driven approach.

**Comment by shentao** [comment]
> I don't think there are any real alternatives of data-driven validation libraries to Vuelidate other than Regle. 

Yeah, I meant more about feature parity so it's easier to understand the relative lift involved with migrating to a different solution and whether it is feasible at all.

> Can I just add that I used Vuelidate for 8 years and it's was always my go-to solution for forms (complex and simple), and by far the best API to handle forms (thanks to Vue reactivity too).

Thanks! Appreciate that! 

I'm actually still using it at the company I'm working at as it fits perfectly into the forms system we created. So from my perspective, the project is still somewhat alive, but due to time limitations, my maintenance work is purely limited to just fixing the problems I encounter personally. And those didn't happen for a while. As our company is moving closer to finally upgrading to Vue 3.x this year, I might spend more time fixing any issues that are specific to Vue 3. 

Hence why I don't want to put a "project is dead" note in the readme. It just feels bad thinking about abandoning the project completely after so much work. Especially now that I see people caring about it. I don't want to promise anything, but I'm tempted to find some time this weekend to work through some of the issues reported. 🤞 

And of course, if more people would want to help spring this whole thing back to life – I'm all open as well!

**Comment by victorgarciaesgi** [comment]
> Hence why I don't want to put a "project is dead" note in the readme

Totally understand, hard to let go!

**Comment by victorgarciaesgi** [comment]
Hey everyone, wanted to let everyone know that my project https://github.com/victorgarciaesgi/regle reached `1.0`! 🎉 

It's using the same API as Vuelidate, but improved in areas such as Typescript, collections and much more, and of course I'm maintaining it regularly as we're using it at my company 👍 

[Take a look at the docs to learn more](https://reglejs.dev)

What Vuelidate brang to the validation world shouldn't die! 💪

**Comment by jacobg** [comment]
Congrats @victorgarciaesgi  and thank you!

**Comment by shentao** [comment]
Congrats @victorgarciaesgi! The docs look stellar. The scoped validation API is a great addition!

**Comment by renatodeleao** [comment]
congrats and thank you @victorgarciaesgi! 🙏

**Comment by martinszeltins** [comment]
@victorgarciaesgi This is great news! Congratulations! Regle is awesome! ❤

**Comment by victorgarciaesgi** [comment]
Thank you all! 💪🏻
